File: module-09-firstvar/appendix/test_reference.py
"""Exactly three acceptance tests for one sorted nonmissing group key."""
import importlib.util
from pathlib import Path
import unittest
import pandas as pd

spec = importlib.util.spec_from_file_location('firstvar_reference', Path(__file__).with_name('reference.py'))
ref = importlib.util.module_from_spec(spec)
spec.loader.exec_module(ref)

class ReferenceTests(unittest.TestCase):
    def check_values(self, rows):
        actual = ref.translate(rows[ref.SOURCE_COLUMNS])
        for key in ['first_flag', 'last_flag']:
            self.assertEqual(actual[key].tolist(), rows['expected_' + key].tolist())

    def test_T1_original_fixture(self):
        rows = ref.load_cases()
        self.check_values(rows.loc[rows.case_id.eq('fixture')])

    def test_T2_single_row_has_both_explicit_flags(self):
        rows = ref.load_cases()
        self.check_values(rows.loc[rows.case_id.eq('edge')])

    def test_T3_preserve_rows_and_integer_flags(self):
        rows = ref.load_cases()
        source = rows.loc[rows.case_id.eq('fixture'), ref.SOURCE_COLUMNS].copy(deep=True)
        source.index = [10, 20, 30, 40, 50, 60]
        before = source.copy(deep=True)
        actual = ref.translate(source)
        self.assertIsNot(actual, source)
        pd.testing.assert_frame_equal(source, before)
        pd.testing.assert_frame_equal(actual[ref.SOURCE_COLUMNS], before)
        for key in ['first_flag', 'last_flag']:
            self.assertTrue(pd.api.types.is_integer_dtype(actual[key]))
            self.assertEqual(actual[key].tolist(), rows.loc[rows.case_id.eq('fixture'), 'expected_' + key].tolist())

if __name__ == '__main__':
    unittest.main(verbosity=2)

File: module-09-firstvar/appendix/test_reference.R
# Exactly three named tests; run with Rscript /absolute/path/test_reference.R.
file_arg <- grep('^--file=', commandArgs(FALSE), value = TRUE)
if (!length(file_arg)) stop('Use Rscript or the documented WebR runner')
here <- dirname(normalizePath(sub('^--file=', '', file_arg[[1]])))
source(file.path(here, 'reference.R'))
cases <- load_cases()
assert_equal <- function(actual, expected) {
  stopifnot(isTRUE(all.equal(actual, expected, check.attributes = FALSE)))
}
check_values <- function(rows) {
  actual <- translate(rows[, source_columns, drop = FALSE])
  assert_equal(actual$first_flag, rows$expected_first_flag)
  assert_equal(actual$last_flag, rows$expected_last_flag)
}
test_T1_original_fixture <- function() {
  check_values(cases[cases$case_id == 'fixture', , drop = FALSE])
}
test_T2_single_row_has_both_explicit_flags <- function() {
  check_values(cases[cases$case_id == 'edge', , drop = FALSE])
}
test_T3_preserve_rows_and_integer_flags <- function() {
  rows <- cases[cases$case_id == 'fixture', , drop = FALSE]
  input <- rows[, source_columns, drop = FALSE]
  rownames(input) <- as.character(seq_len(nrow(input)) * 10L)
  before <- input
  actual <- translate(input)
  stopifnot(identical(input, before), identical(actual[source_columns], before))
  stopifnot(is.integer(actual$first_flag), is.integer(actual$last_flag))
  assert_equal(actual$first_flag, rows$expected_first_flag)
  assert_equal(actual$last_flag, rows$expected_last_flag)
}

test_T1_original_fixture(); cat("PASS T1 original_fixture\n")
test_T2_single_row_has_both_explicit_flags(); cat("PASS T2 single_row_has_both_explicit_flags\n")
test_T3_preserve_rows_and_integer_flags(); cat("PASS T3 preserve_rows_and_integer_flags\n")
